### PR TITLE
Implement Sided Components

### PR DESCRIPTION
**Pull request by ExE-Boss**
This PR implements sided components. This allows for Blocks to have specific components on one side only.
Required for Minecraft 1.9+ wrappers for compatibility with Forge’s Capabilities.

### CODE REVIEW

File: minecraft/1.7/src/main/java/nova/core/wrapper/mc/forge/v17/asm/lib/ComponentInjector.java [hunk 1]
```diff
@@ -60,7 +60,8 @@ public ComponentInjector(Class<T> baseClazz) {
 		this.baseClazz = baseClazz;
 	}
 
-	public synchronized T inject(ComponentProvider provider, Class<?>[] typeArgs, Object[] args) {
+	@SuppressWarnings("rawtypes")
+	public synchronized T inject(ComponentProvider<? extends ComponentProvider<? extends ComponentProvider.ComponentMap>.ComponentMap> provider, Class<?>[] typeArgs, Object[] args) {
```
Comment (RX14) [inline minecraft/1.7/src/main/java/nova/core/wrapper/mc/forge/v17/asm/lib/ComponentInjector.java]: These types are really ugly. Is there not a better way to handle this?

Comment (ExE-Boss) [inline minecraft/1.7/src/main/java/nova/core/wrapper/mc/forge/v17/asm/lib/ComponentInjector.java]: I tried, but sadly, there isn't.

Comment (ExE-Boss) [inline minecraft/1.7/src/main/java/nova/core/wrapper/mc/forge/v17/asm/lib/ComponentInjector.java]: This is what happens when you have a non-static class that you pass as a type to the class that contains said non-static class

Comment (RX14) [inline minecraft/1.7/src/main/java/nova/core/wrapper/mc/forge/v17/asm/lib/ComponentInjector.java]: Make the class static?

File: src/main/java/nova/core/component/SidedComponentMap.java [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+	}
+
+	/**
+	 * Checks if a component type exists in this provider.
+	 * @param componentType the component type to check.
+	 * @param direction The direction to check.
+	 * @return true if the component exists on the provider.
+	 */
+	public final boolean has(Class<?> componentType, Direction direction) {
+		return getComponents(direction).has(componentType) | this.has(componentType);
```
Comment (RX14) [inline src/main/java/nova/core/component/SidedComponentMap.java]: `||`?

File: src/main/java/nova/core/component/ComponentProvider.java [hunk 3]
```diff
@@ -60,6 +48,7 @@ public ComponentProvider() {
 	@SuppressWarnings("unchecked")
 	public <C extends ComponentMap> ComponentProvider(Class<C> componentsClass) {
 		try {
+			// TODO Figure out how to dependency injection with this without it crashing
```
Comment (RX14) [inline src/main/java/nova/core/component/ComponentProvider.java]: I'd rather not merge something which crashes.

Comment (ExE-Boss) [inline src/main/java/nova/core/component/ComponentProvider.java]: It's dependency injection that causes the crash on this line, not reflection.
It works fine like this

File: src/main/java/nova/core/component/SidedComponentMap.java (hunk 2 above)
Comment (ExE-Boss) [inline src/main/java/nova/core/component/SidedComponentMap.java]: It doesn't really matter with booleans

File: src/main/java/nova/core/component/SidedComponentMap.java [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+	}
+
+	public final Map<Direction, ComponentMap> getComponentsForDirections() {
+		return Arrays.stream(Direction.values()).collect(Collectors.toMap(Function.identity(), this::getComponents));
+	}
+
+	public final ComponentMap getComponents(Direction direction) {
+		switch (direction) {
+			case UP:
+				return up;
```
Comment (Victorious3) [inline src/main/java/nova/core/component/SidedComponentMap.java]: An array, maybe?

**Pull request merged**